### PR TITLE
register template part variations server side.

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
## Description

This explores refactoring how we define template part block variations.  Instead of the current approach (subscribing to the store and waiting various conditions to register the variations), here we register the variations on the server side and patch the `isActive` and `icon` properties on the JS side with a filter.  I am not sure whether this approach makes more sense than the last, but explored it while trying to debug issues regarding variations not showing up for some people (after more investigation #31772 fixes this issue without this refactor).

## How has this been tested?

run `npm run build` and verify variations show up and work as expected.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/template-part/variations.js [hunk 1]
```diff
@@ -2,41 +2,16 @@
...
 
-const unsubscribe = subscribe( () => {
-	const definedVariations = select(
-		editorStore
-	).__experimentalGetDefaultTemplatePartAreas();
-
-	if ( ! definedVariations?.length ) {
-		return;
+export function enhanceTemplatePartVariations( settings, name ) {
+	if ( name !== 'core/template-part' ) {
```
Comment (ntsekouras) [inline packages/block-library/src/template-part/variations.js:30]: Nitpick: we can add the check here for `! settings.variations`.

File: packages/block-library/src/template-part/variations.js [hunk 2]
```diff
@@ -51,10 +26,21 @@ const unsubscribe = subscribe( () => {
...
-	dispatch( blocksStore ).addBlockVariations(
-		'core/template-part',
-		variations
-	);
-} );
+		const variations = settings.variations.map( ( variation ) => {
+			return {
+				...variation,
+				...( ! variation.isActive && { isActive } ),
+				...( typeof variation.icon === 'string' && {
```
Comment (ntsekouras) [inline packages/block-library/src/template-part/variations.js:61]: Should we check `icons[ variation.icon ] || set the TP default icon`? The `typeof === 'string'` can be omitted. 

Also related to that, should we just create a mapper object that would contain the `header` and `footer` icons to avoid importing all the icons and would default to `layout`? --cc @gziolo

**Review by ntsekouras — COMMENTED** [review]
I think this is a more consistent way of handling the variations compared to https://github.com/WordPress/gutenberg/pull/31772. I vote for this one 😄

**Comment by jameskoster** [comment]
FWIW, this one isn't working for me whereas 31772 is.

**Comment by ntsekouras** [comment]
>FWIW, this one isn't working for me whereas 31772 is.

That's odd 🤔 . Have you run `npm run build` in this branch and the variations still not showing?

**Comment by jameskoster** [comment]
Yup, unfortunately that is the case.

**Comment by Copons** [comment]
This is not working for me as well.
I can pinpoint where it fails, but I'm not sure why. 🤔 

The filters executes correctly, so the build side effect works as expected, but `settings.variations` is undefined.
Screenshot 2021-05-13 at 12 58 43

Though, the block seems to be registered correctly in the PHP side, and `build_template_part_block_variations()` does indeed return the variations as expected.
So I guess something happens between the PHP registration and the JS counterpart. 🤔 

Apparently, the Navigation Link block is the same, but I have no idea if that's intended (as in: it's not supposed to have variations added this way?). cc @gwwar 


EDIT: I've tested this both as `npm run build` and `npm run dev` with the same outcome.

EDIT 2: either way, I like this approach much more than #31772! ✨

**Comment by Addison-Stavlo** [comment]
oof, so we like the approach that doesn't seem to be working for some folks.  Im not sure why this filter would work for some and not others. 😕

**Comment by Addison-Stavlo** [comment]
> So I guess something happens between the PHP registration and the JS counterpart. 🤔
> Apparently, the Navigation Link block is the same,

It looks like the Navigation Link gets around this issue by having fallback variations defined in JS to register in the case that `settings.variations` is not defined.  Im not sure the reasoning there, but that doesn't seem like a good approach here.  If our goal is for users to have access to the variations as defined on the server side, but half of the users are only seeing the fallback variations that are hardcoded then that we missed the mark on the server side definitions in the first place.

Since this filter is not behaving reliably and consistent, but the subscribe approach is, id say we stick with that at least for the bug fix.  I don't see whats wrong with a `subscribe` that `unsubscribes` itself as soon as the store is initialized. 
 I say we look at https://github.com/WordPress/gutenberg/pull/31772 for the bug fix, it just ensures the code that is already there and working actually runs with the production build.  We can keep investigating this approach and hopefully find a way to get it working consistently.

File: packages/block-library/src/template-part/variations.js (hunk 2 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/variations.js:61]: That all makes sense.  

> typeof === 'string' can be omitted.

I was wondering if maybe there was a way someone could have registered it with an actual icon instead of an icon name, and to just double check if it was a simple string at this point before trying to overwrite it.

**Comment by Addison-Stavlo** [comment]
>. So I guess something happens between the PHP registration and the JS counterpart. 🤔
 
 @Copons  - out of curiosity... does it start working for you if you specify an earlier priority on the `add_action( 'init', 'register_block_core_template_part' );` in index.php?

**Comment by Copons** [comment]
> @Copons - out of curiosity... does it start working for you if you specify an earlier priority on the `add_action( 'init', 'register_block_core_template_part' );` in index.php?

@Addison-Stavlo Tried with `add_action( 'init', 'register_block_core_template_part', 1 );`, and it still doesn't work. ☹️ 

I'm cool with going with the subscription approach, but I'd first ask the Navigation Link folks if they have any pointers.

What I meant when I said that it's not working as well, is that NavLink's original `settings.variations` (which I assume is what is returned by the block registration in PHP?) is undefined **before** reaching the fallbacks.
So basically it's the very same issue, just NavLink then fills the variations with fallbacks.

Again, I'm not familiar with that block, but I think it's worthwhile to check if it's broken too, as the cause might be the same that breaks this for some of us.

**Comment by Addison-Stavlo** [comment]
>Again, I'm not familiar with that block, but I think it's worthwhile to check if it's broken too

100% agree

**Comment by gwwar** [comment]
See example in https://github.com/WordPress/gutenberg/pull/29095

Server variation support landed in WP 5.8, so there's still a need for JS fallback variations with WP 5.7 and below, until Gutenberg requires at least WP 5.8.

https://github.com/WordPress/gutenberg/blob/815b767a46f03513143b15cbd200f213306f7d18/gutenberg.php#L6

**Comment by Copons** [comment]
> See example in #29095
> 
> Server variation support landed in WP 5.8, so there's still a need for JS fallback variations with WP 5.7 and below, until Gutenberg requires at least WP 5.8.
> 
> https://github.com/WordPress/gutenberg/blob/815b767a46f03513143b15cbd200f213306f7d18/gutenberg.php#L6

Interesting!

I'm currently running 5.8-alpha-50507 on my local wp-env, so I guess I'm not seeing the variations as the patch containing the server support is not included yet in my version 🤔

**Comment by Addison-Stavlo** [comment]
This should work now with fallbacks for folks on earlier versions that don't support the server side registration.

File: packages/block-library/src/template-part/variations.js (hunk 2 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/variations.js:61]: >  should we just create a mapper object that would contain the header and footer icons to avoid importing all the icons and would default to layout?

Yeah, I'l go ahead and limit this to just importing the icons we expect for template parts at this time (header, footer, sidebar, and layout).

It would be neat if in the future we could enable registering custom icons with a template part area.  Its no requirement at this point, but if instead of defining `'icon' => 'icon_name'` on the php side we could define `'icon' => $some_icon`.  But I'm currently drawing a blank on how one might go about that.

Comment (gwwar) [reply packages/block-library/src/template-part/variations.js:61]: I was planning on exploring allowing icons to be set with svg strings, but the tricky part is finding a solution that will also work with react native

**Review by Copons — APPROVED** [review]
This is working well for me now! ✨ 

@jameskoster would you mind trying again just to be 100% sure?

**Comment by jameskoster** [comment]
It works 🥳

File: packages/block-library/src/template-part/variations.js (hunk 2 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/variations.js:61]: @gwwar that would be neat!  Im not too familiar with native's limitations there.
